File: lib/iriq/trace.rb
module Iriq
  # Produces an annotated trace explaining how an identifier got
  # normalized — segment by segment, with notes for each non-obvious
  # transformation (currency upcase, IP umbrella, hint suppression,
  # canonical date, param-name lift, etc.).
  #
  #   Trace.for("https://shop.com/pricing/usd?currency=eur")
  #   # => {
  #   #      input: "...",
  #   #      normalized: "https://shop.com/pricing/USD?currency=EUR",
  #   #      scheme: "https", host: "shop.com",
  #   #      path:  [...per-segment rows...],
  #   #      query: [...per-param rows...],
  #   #    }
  #
  # Each row is `{ value, type, output, notes }` for path entries and
  # `{ name, value, type, output, notes }` for query entries. The string
  # notes are rendered from structured Iriq::Evidence::Record values;
  # callers that want the structured form can use Trace.evidence_for.
  module Trace
    module_function

    HINT_NOTE_TEMPLATE = "semantic type — surfaced as {%s}, not {%s}".freeze

    # Render-ready Trace output. The public format consumers depend on.
    def for(input, classifier: SegmentClassifier::DEFAULT, hints: true)
      iri = coerce(input)
      normalized = Normalizer.normalize_identifier(iri, classifier: classifier, hints: hints)

      out = {
        input:      iri.canonical,
        normalized: normalized,
        scheme:     iri.scheme,
        host:       iri.host,
      }
      out[:port] = iri.port if iri.port

      if iri.urn?
        out[:path] = urn_rows(iri, classifier, hints)
      else
        out[:path]  = path_rows(iri.path_segments, classifier, hints)
        if iri.query_params && !iri.query_params.empty?
          out[:query] = query_rows(iri.query_params, classifier)
        end
      end

      out
    end

    # Structured Evidence list for `input`. Each segment + query param
    # contributes one classification Evidence plus zero or more
    # transformation Evidence records (canonical date, IP umbrella
    # collapse, param-name hint, hint suppression).
    #
    # Position + Cluster Evidence are not emitted here — they belong to
    # corpus-informed trace (Corpus#trace), which a follow-up step lands.
    def evidence_for(input, classifier: SegmentClassifier::DEFAULT, hints: true)
      iri      = coerce(input)
      records  = []
      segments = iri.urn? ? urn_parts(iri) : (iri.path_segments || [])
      entries  = SegmentHints.derive(segments, classifier)

      entries.each_with_index do |entry, i|
        records.concat(segment_evidence(entry, segments, i, classifier, hints))
      end

      if !iri.urn? && iri.query_params && !iri.query_params.empty?
        iri.query_params.keys.sort.each do |k|
          records.concat(query_param_evidence(k, iri.query_params[k].to_s, classifier))
        end
      end

      records
    end

    # ── Evidence builders ────────────────────────────────────────────────

    def segment_evidence(entry, segments, idx, classifier, hints)
      records = []

      records << Evidence.segment(
        index: idx, value: entry[:value],
        source:  :recognizer,
        payload: { type: entry[:type], variable: entry[:variable], hint: entry[:hint] },
      )

      if entry[:variable]
        if entry[:type] == :date && (canon = SegmentClassifier.canonical_date(entry[:value]))
          if canon != entry[:value]
            records << Evidence.segment(
              index: idx, value: entry[:value],
              source:  :policy,
              payload: { rule: :canonical_date, before: entry[:value], after: canon },
              notes:   ["canonical date (#{entry[:value]} → #{canon})"],
            )
          end
        elsif entry[:type] == :currency && (canon = SegmentClassifier.canonical_currency(entry[:value]))
          if canon != entry[:value]
            records << Evidence.segment(
              index: idx, value: entry[:value],
              source:  :policy,
              payload: { rule: :canonical_currency, before: entry[:value], after: canon },
              notes:   ["currency upcase (#{entry[:value]} → #{canon})"],
            )
          end
        else
          extra = placeholder_decoration_evidence(entry, segments, idx, classifier, hints)
          records.concat(extra)
        end
      end

      records
    end

    def placeholder_decoration_evidence(entry, segments, idx, classifier, hints)
      out  = []
      type = entry[:type]

      if type == :ipv4 || type == :ipv6
        out << Evidence.segment(
          index: idx, value: entry[:value],
          source:  :policy,
          payload: { rule: :ip_umbrella_collapse, from: type, to: :ip },
          notes:   ["ip umbrella collapse (#{type} → ip)"],
        )
      end

      if hints && entry[:hint].nil? && !SegmentHints::HINT_ELIGIBLE_TYPES.include?(type)
        if (would_be = would_be_hint(segments, idx, type, classifier))
          display = SegmentClassifier.display_type(type)
          out << Evidence.segment(
            index: idx, value: entry[:value],
            source:  :neighbor,
            payload: { rule: :hint_suppression, surfaced: display, would_be: would_be, semantic_type: type },
            notes:   [format(HINT_NOTE_TEMPLATE, display, would_be)],
          )
        end
      end

      out
    end

    def query_param_evidence(name, value, classifier)
      records   = []
      base_type = classifier.classify(value)
      effective = base_type

      if (hint = SegmentClassifier.param_name_hint(name, base_type))
        effective = hint
        records << Evidence.segment(
          index: name, value: value,
          source:  :neighbor,
          payload: { rule: :param_name_hint, name: name, before: base_type, after: hint },
          notes:   ["param-name hint (`#{name}=`) lifted #{base_type} → #{hint}"],
        )
      end

      records << Evidence.segment(
        index: name, value: value,
        source:  :recognizer,
        payload: { type: effective, variable: SegmentClassifier::DEFAULT.variable?(effective) },
      )

      if effective == :date && (canon = SegmentClassifier.canonical_date(value))
        if canon != value
          records << Evidence.segment(
            index: name, value: value,
            source:  :policy,
            payload: { rule: :canonical_date, before: value, after: canon },
            notes:   ["canonical date (#{value} → #{canon})"],
          )
        end
      elsif effective == :currency && (canon = SegmentClassifier.canonical_currency(value))
        if canon != value
          records << Evidence.segment(
            index: name, value: value,
            source:  :policy,
            payload: { rule: :canonical_currency, before: value, after: canon },
            notes:   ["currency upcase (#{value} → #{canon})"],
          )
        end
      elsif effective == :ipv4 || effective == :ipv6
        records << Evidence.segment(
          index: name, value: value,
          source:  :policy,
          payload: { rule: :ip_umbrella_collapse, from: effective, to: :ip },
          notes:   ["ip umbrella collapse (#{effective} → ip)"],
        )
      end

      records
    end

    # ── View rendering (Evidence → Trace.for hash) ───────────────────────

    def path_rows(segments, classifier, hints)
      return [] if segments.nil? || segments.empty?

      entries = SegmentHints.derive(segments, classifier)
      entries.each_with_index.map do |entry, i|
        ev = segment_evidence(entry, segments, i, classifier, hints)
        render_segment_row(entry, ev, hints)
      end
    end

    def urn_rows(iri, classifier, hints)
      parts = urn_parts(iri)
      return [] if parts.empty?

      entries = SegmentHints.derive(parts, classifier)
      entries.each_with_index.map do |entry, i|
        ev = segment_evidence(entry, parts, i, classifier, hints)
        render_segment_row(entry, ev, hints)
      end
    end

    def query_rows(params, classifier)
      params.keys.sort.map do |k|
        v  = params[k].to_s
        ev = query_param_evidence(k, v, classifier)
        render_query_row(k, v, ev)
      end
    end

    def render_segment_row(entry, evidence, hints)
      notes = collect_notes(evidence)
      value = entry[:value]
      type  = entry[:type]

      return { value: value, type: type, output: value, notes: notes } unless entry[:variable]

      canon_policy = find_payload(evidence, :canonical_date) || find_payload(evidence, :canonical_currency)
      if canon_policy
        return { value: value, type: type, output: canon_policy[:after], notes: notes }
      end

      placeholder = hints && entry[:hint] ? entry[:hint].to_s : SegmentClassifier.display_type(type).to_s
      { value: value, type: type, output: "{#{placeholder}}", notes: notes }
    end

    def render_query_row(name, value, evidence)
      notes        = collect_notes(evidence)
      cls          = find_evidence(evidence, source: :recognizer)
      effective    = cls ? cls.payload[:type] : SegmentClassifier::DEFAULT.classify(value)
      canon_policy = find_payload(evidence, :canonical_date) || find_payload(evidence, :canonical_currency)

      output =
        if canon_policy
          canon_policy[:after]
        elsif SegmentClassifier::DEFAULT.variable?(effective)
          "{#{SegmentClassifier.display_type(effective)}}"
        else
          value
        end

      { name: name, value: value, type: effective, output: output, notes: notes }
    end

    # ── Helpers ──────────────────────────────────────────────────────────

    def coerce(input)
      input.is_a?(Identifier) ? input : Parser.parse(input)
    end

    def urn_parts(iri)
      return [] unless iri.nss
      iri.nss.include?(":") ? iri.nss.split(":", 2) : [iri.nss]
    end

    def collect_notes(evidence)
      evidence.flat_map(&:notes)
    end

    def find_evidence(evidence, source:)
      evidence.find { |r| r.source == source }
    end

    def find_payload(evidence, rule)
      r = evidence.find { |e| e.source == :policy && e.payload[:rule] == rule }
      r&.payload
    end

    def would_be_hint(segments, idx, type, classifier)
      return nil if idx.zero?

      prev = segments[idx - 1]
      return nil unless classifier.classify(prev) == :literal

      base   = Inflector.singularize(prev)
      suffix = type == :uuid ? "_uuid" : "_id"
      "#{base}#{suffix}"
    end
  end
end

File: rust/iriq/src/trace.rs
use crate::classifier::{
    canonical_currency, canonical_date, display_type, param_name_hint, SegmentClassifier,
    SegmentType, DEFAULT_CLASSIFIER,
};
use crate::errors::ParseError;
use crate::hints::{derive_hints, hint_eligible_types, SegmentHint};
use crate::identifier::Identifier;
use crate::inflector::singularize;
use crate::normalizer::normalize_identifier;
use crate::parser::parse;

#[derive(Debug, Clone, serde::Serialize)]
pub struct TraceRow {
    #[serde(skip_serializing_if = "String::is_empty")]
    pub name: String,
    pub value: String,
    #[serde(rename = "type")]
    pub ty: SegmentType,
    pub output: String,
    pub notes: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct TraceResult {
    pub input: String,
    pub normalized: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub scheme: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub host: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub port: Option<u64>,
    pub path: Vec<TraceRow>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub query: Vec<TraceRow>,
}

pub fn trace(input: &str) -> Result<TraceResult, ParseError> {
    let iri = parse(input)?;
    Ok(trace_identifier(&iri, &DEFAULT_CLASSIFIER, true))
}

pub fn trace_identifier(iri: &Identifier, c: &SegmentClassifier, hints: bool) -> TraceResult {
    let mut out = TraceResult {
        input: iri.canonical(),
        normalized: normalize_identifier(iri, c, hints),
        scheme: iri.scheme.clone(),
        host: iri.host.clone(),
        port: iri.port,
        path: Vec::new(),
        query: Vec::new(),
    };
    if iri.is_urn() {
        let segments = urn_parts(iri);
        out.path = trace_path(&segments, c, hints);
        return out;
    }
    out.path = trace_path(&iri.path_segments, c, hints);
    if !iri.query_params.is_empty() {
        out.query = trace_query(iri, c);
    }
    out
}

fn urn_parts(iri: &Identifier) -> Vec<String> {
    if iri.nss.is_empty() {
        return Vec::new();
    }
    if let Some((ns, val)) = iri.nss.split_once(':') {
        vec![ns.to_string(), val.to_string()]
    } else {
        vec![iri.nss.clone()]
    }
}

fn trace_path(segments: &[String], c: &SegmentClassifier, hints: bool) -> Vec<TraceRow> {
    if segments.is_empty() {
        return Vec::new();
    }
    let entries = derive_hints(segments, c);
    entries
        .iter()
        .enumerate()
        .map(|(i, e)| render_segment_row(e, segments, i, c, hints))
        .collect()
}

fn render_segment_row(
    entry: &SegmentHint,
    segments: &[String],
    idx: usize,
    c: &SegmentClassifier,
    hints: bool,
) -> TraceRow {
    let mut notes = Vec::new();
    if !entry.variable {
        return TraceRow {
            name: String::new(),
            value: entry.value.clone(),
            ty: entry.ty,
            output: entry.value.clone(),
            notes,
        };
    }

    // Canonical date / currency take precedence over placeholder rendering.
    if entry.ty == SegmentType::Date {
        if let Some(canon) = canonical_date(&entry.value) {
            if canon != entry.value {
                notes.push(format!("canonical date ({} → {})", entry.value, canon));
            }
            return TraceRow {
                name: String::new(),
                value: entry.value.clone(),
                ty: entry.ty,
                output: canon,
                notes,
            };
        }
    }
    if entry.ty == SegmentType::Currency {
        if let Some(canon) = canonical_currency(&entry.value) {
            if canon != entry.value {
                notes.push(format!("currency upcase ({} → {})", entry.value, canon));
            }
            return TraceRow {
                name: String::new(),
                value: entry.value.clone(),
                ty: entry.ty,
                output: canon,
                notes,
            };
        }
    }

    // IP umbrella collapse note
    if entry.ty == SegmentType::Ipv4 || entry.ty == SegmentType::Ipv6 {
        notes.push(format!("ip umbrella collapse ({} → ip)", entry.ty));
    }

    // Hint suppression note for semantic types.
    if hints && entry.hint.is_empty() && !hint_eligible_types().contains(&entry.ty) {
        if let Some(would) = would_be_hint(segments, idx, entry.ty, c) {
            let display = display_type(entry.ty);
            notes.push(format!(
                "semantic type — surfaced as {{{}}}, not {{{}}}",
                display, would
            ));
        }
    }

    let output = if hints && !entry.hint.is_empty() {
        format!("{{{}}}", entry.hint)
    } else {
        format!("{{{}}}", display_type(entry.ty))
    };
    TraceRow {
        name: String::new(),
        value: entry.value.clone(),
        ty: entry.ty,
        output,
        notes,
    }
}

fn would_be_hint(
    segments: &[String],
    idx: usize,
    t: SegmentType,
    c: &SegmentClassifier,
) -> Option<String> {
    if idx == 0 {
        return None;
    }
    let prev = &segments[idx - 1];
    if c.classify(prev) != SegmentType::Literal {
        return None;
    }
    let base = singularize(prev);
    let suffix = if t == SegmentType::Uuid {
        "_uuid"
    } else {
        "_id"
    };
    Some(format!("{}{}", base, suffix))
}

fn trace_query(iri: &Identifier, c: &SegmentClassifier) -> Vec<TraceRow> {
    let mut keys = iri.query_params.keys();
    keys.sort();
    let mut rows = Vec::with_capacity(keys.len());
    for k in keys {
        let v = iri.query_params.get(&k).unwrap_or("").to_string();
        rows.push(render_query_row(&k, &v, c));
    }
    rows
}

fn render_query_row(name: &str, value: &str, c: &SegmentClassifier) -> TraceRow {
    let mut notes = Vec::new();
    let base = c.classify(value);
    let effective = if let Some(h) = param_name_hint(name, base) {
        notes.push(format!(
            "param-name hint (`{}=`) lifted {} → {}",
            name, base, h
        ));
        h
    } else {
        base
    };

    let output = match effective {
        SegmentType::Date => {
            if let Some(canon) = canonical_date(value) {
                if canon != value {
                    notes.push(format!("canonical date ({} → {})", value, canon));
                }
                canon
            } else if c.variable(effective) {
                format!("{{{}}}", display_type(effective))
            } else {
                value.to_string()
            }
        }
        SegmentType::Currency => {
            if let Some(canon) = canonical_currency(value) {
                if canon != value {
                    notes.push(format!("currency upcase ({} → {})", value, canon));
                }
                canon
            } else if c.variable(effective) {
                format!("{{{}}}", display_type(effective))
            } else {
                value.to_string()
            }
        }
        SegmentType::Ipv4 | SegmentType::Ipv6 => {
            notes.push(format!("ip umbrella collapse ({} → ip)", effective));
            format!("{{{}}}", display_type(effective))
        }
        _ => {
            if c.variable(effective) {
                format!("{{{}}}", display_type(effective))
            } else {
                value.to_string()
            }
        }
    };

    TraceRow {
        name: name.to_string(),
        value: value.to_string(),
        ty: effective,
        output,
        notes,
    }
}
